src/Plugins/Loader/AssemblyLoadContextBuilder.cs

```diff
@@ -148,21 +150,46 @@ public AssemblyLoadContextBuilder PreferLoadContextAssembly(AssemblyName assembl
         /// <returns>The builder.</returns>
         public AssemblyLoadContextBuilder PreferDefaultLoadContextAssembly(AssemblyName assemblyName)
         {
-            if (assemblyName.Name == null || _defaultAssemblies.Contains(assemblyName.Name))
+            // Lazy loaded references have dependencies resolved as they are loaded inside the actual Load Context.
+            if (_lazyLoadReferences)
             {
-                // base cases
+                if (assemblyName.Name != null && !_defaultAssemblies.Contains(assemblyName.Name))
+                {
+                    _defaultAssemblies.Add(assemblyName.Name);
+                    var assembly = _defaultLoadContext.LoadFromAssemblyName(assemblyName);
+                    foreach (var reference in assembly.GetReferencedAssemblies())
+                    {
+                        if (reference.Name != null)
+                        {
+                            _defaultAssemblies.Add(reference.Name);
+                        }
+                    }
+                }
+
                 return this;
             }
+            
+            var names = new Queue<AssemblyName>();
+            names.Enqueue(assemblyName);
+            while (names.TryDequeue(out var name))
+            {
+                if (name.Name == null || _defaultAssemblies.Contains(name.Name))
+                {
+                    // base cases
+                    continue;
+                }
 
-            _defaultAssemblies.Add(assemblyName.Name);
+                _defaultAssemblies.Add(name.Name);
 
-            // Recursively load and find all dependencies of default assemblies.
-            // This sacrifices some performance for determinism in how transitive
-            // dependencies will be shared between host and plugin.
-            var assembly = _defaultLoadContext.LoadFromAssemblyName(assemblyName);
-            foreach (var reference in assembly.GetReferencedAssemblies())
-            {
-                PreferDefaultLoadContextAssembly(reference);
+                // Load and find all dependencies of default assemblies.
+                // This sacrifices some performance for determinism in how transitive
+                // dependencies will be shared between host and plugin.
+                var assembly = _defaultLoadContext.LoadFromAssemblyName(name);
+
+                foreach (var reference in assembly.GetReferencedAssemblies())
+                {
+                    names.Enqueue(reference);
+                }
             }
```

Comment:
What is the purpose of this refactor?

Reply:
I did it because the original method was recursive.

I wasn't really sure whether you would have preferred to (from a style perspective) to have a separate internal method for the non-lazy implementation. Personally, I just thought the idea of checking a branch that would always have been false in the beginning seemed a bit silly to me.

Doing it this way is also technically faster due to less branches/jumps; albeit by such a negligible non-perceptible amount it's not worth caring about.